### ISSUE TITLE
How to get the IO address by reading the log files generated by write_iolog

### ISSUE THREAD

**Issue by wangyihanlarry** [axboe/fio#1135]
would you please give me a bit more details on the output of the file that generate by appending the switch write_iolog, I got the following outputs and I am wondering how to convert the third column into the address which services the IO,

/dev/sdhu add
/dev/sdhu open
/dev/sdhu read 129521664 2048 # E.g. how to get this into the address with 1M unit?
/dev/sdhu read 1589760000 10240
/dev/sdhu read 1810721792 12288
/dev/sdhu write 1009347584 3072
/dev/sdhu read 863098880 5120
/dev/sdhu read 1804161024 14336
/dev/sdhu write 817105920 1024
/dev/sdhu write 759179264 6144
/dev/sdhu write 1960665088 8192
/dev/sdhu read 143502336 1024
/dev/sdhu write 1836585984 8192
/dev/sdhu read 1398337536 15360
/dev/sdhu write 1994024960 4096
/dev/sdhu read 530178048 2048
/dev/sdhu read 842858496 3072

I am asking this as I am not able to divide the 129521664 by any of 512 or 1024

**Comment by sitsofe** [comment]
The third column numbers are the block offset requested in bytes (as stated in https://fio.readthedocs.io/en/latest/fio_doc.html#trace-file-format-v2 ). All those number look evenly dividable by 512 without remainder:
```
python3
>>> l = [129521664, 1589760000, 1810721792, 1009347584, 863098880, 1804161024, 817105920, 759179264, 1960665088, 143502336, 1836585984, 1398337536, 1994024960, 530178048, 842858496]
>>> ["Even" if i % 512.0 == 0 else "Remainder" for i in l]
['Even', 'Even', 'Even', 'Even', 'Even', 'Even', 'Even', 'Even', 'Even', 'Even', 'Even', 'Even', 'Even', 'Even', 'Even']
```

so I don't think your final statement is correct (`129521664 % 512 == 0`).

> address which services the IO

I'd guess you need to look up how to find out the logical block size of your disk to answer this one.

**Comment by wangyihanlarry** [comment]
Thanks for your quick response! I will try it out and get back to you when there is any more issues, another question is will the '0' literately considered as empty, saying that I am writing all the 0 for the following item.

/dev/sdhu read 1589760000 10240

**Comment by sitsofe** [comment]
> will the '0' literately considered as empty

I'm afraid I don't understand what you mean when you say "empty" means in this context... Could you clarify further? Also how do you know zeroes are being written (do you know that fio defaults to writing random data? See the fio documentation about buffers and memory - https://fio.readthedocs.io/en/latest/fio_doc.html#buffers-and-memory ?)

**Comment by sitsofe** [comment]
Setting to needreporterinfo pending a reply from @wangyihanlarry .

**Comment by wangyihanlarry** [comment]
sorry for the confusion, I have figured out the answer myself, even if we write the 0 into the block, the device will also allocate the space holding for it so the bitmap should be 1 in this case

However, I have met any issues need to address, I attached my fio profile and my programs to generate the bitmap into the thread, here is the details steps I have tried in below

1st, I create a 1gb volume and attach it into the IO host, after that I send the FIO with following profile:

[fio_profile.fio.txt](https://github.com/axboe/fio/files/5612370/fio_profile.fio.txt)

2nd, After the completion the IO, I got the following io write log as you mentioned, also in attachment:

[testlarrry.txt](https://github.com/axboe/fio/files/5612372/testlarrry.txt)
here if you look into the line 507, you will find the following output, this offset is beyond the 1GB
`/dev/sdfe write 62952243113008640 90112
`
3rd, also below is the code I used to generate the bitmap in 1MB basis.
`import io
import re

temps = {}
bit_map = {}
with io.open('/root/1020/testlarrry', 'r') as temp:
    for line in temp.readlines():
        # hit = re.search(r"({})\s+=\s?(\S+)#?".format(attribute), element_configruation)
        hit = re.search(r"/dev/(\S+)\s+write\s+(\d+)\s+(\d+)", line)
        if hit:
            dev_name = hit.group(1)
            offset = hit.group(2)
            length = hit.group(3)
            start_in_mb = int(offset)/1024/1024
            end_in_mb = (int(offset)+int(length))/1024/1024
            if (end_in_mb > 1024):
                print(start_in_mb)
                print(end_in_mb)
                print(offset)
                print(length)
            temps.setdefault(dev_name, {start_in_mb: end_in_mb})
            temps[dev_name].setdefault(start_in_mb, end_in_mb)

for key in temps.keys():
    # TODO: Key is the device number, we need firstly featch the size of device and then intitialze the bitmap list
    size = 1024
    init_bitmap = [0] * size
    enrtys = temps[key].keys()
    # TODO: walk through the entry to update the init_bitmap
    for entry in enrtys:
        end = temps[key][entry]
        init_bitmap[entry:end] = [1] * ((end - entry) + 1)
    bit_map.setdefault(key, init_bitmap)

pass
`

**Comment by sitsofe** [comment]
WARNING: You have moved this "issue" away from your initial title ("how do I get the address?") twice...

First up, thank you for posting which version of fio you were using and the options you were running with. Can you make sure that you [**reproduce your problems with the latest fio**](https://github.com/axboe/fio/blob/fio-3.24/REPORTING-BUGS#L6) (see https://github.com/axboe/fio/releases for which version we're up to). A **lot** of things have been fixed since fio-3.1 and some of those solve your issue...

Secondly, as this isn't so much an issue in fio and more of a "How do I/Why is?" discussion it is better asked on the [fio mailing list](https://github.com/axboe/fio/blob/fio-3.24/README#L58) (note that the list only accepts **plain text emails** so make sure you set your mail program to only send that type of mail to the list). I'll close this issue here and we can pick it up on the list.

Thanks!

PS: "all zeroes" can be a special value for storage and subject to things like detection/compression so all sorts of special behaviour can kick in that make results unrealistic - real data is not usually all zeroes beyond initial formatting.
